doc: Cluster to advanced cluster migration guide (#2451)
mongodb/terraform-provider-mongodbatlas: docs/guides/cluster-to-advanced-cluster-migration-guide.md

# Cluster to Advanced Cluster Migration Guide

**Objective**: Guide users to replace the `mongodbatlas_cluster` resource with the `mongodbatlas_advanced_cluster` resource.

> -> **NOTE:** This guide focus on the resource migration as the data source migration is only requiring a resource_type change from `data.mongodbatlas_cluster` to `data.mongodbatlas_advanced_cluster`.  However, pay attention to the [output changes.](#output-changes)

Reviewer: ```suggestion
-> **NOTE:** This guide focuses on the resource migration, as the data source migration only requires a `resource_type` change from `data.mongodbatlas_cluster` to `data.mongodbatlas_advanced_cluster`.  However, pay attention to the [output changes.](#output-changes)
```

Author: Thank you! Updated

## Main Changes Between `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`

> Key differences in the configuration:

Reviewer: Nit: this list heading seems a bit redundant (basically the same as the section heading above?)

Author: Agree. Removed

> 1. Replication Spec Configuration: Supports different node types (electable, analytics, read_only) where hardware configuration can differ between node types.

Reviewer: It might make sense to list these in the same order that they appear in the TF file itself. Also, in the code example, it might be nice to add commented out block headings like you have for "Provider Settings" just to make reference simple.

Author: Changed in 288f7f76a58ef30cf833e0adbb1e9543b719d138
There are lines that ends with a comment referencing the number and the header. In Example 1 they follow the numeration
WDYT?

2. Provider Settings: Moved from the top level to the replication spec allowing you to create multi-cloud clusters.
3. Auto Scaling: Moved from the top level to the replication spec allowing you to scale replication specs individually.
4. Backup Configuration: Renamed from `cloud_backup` to `backup_enabled`.
5. See also ["Migration to new sharding schema and leveraging Independent Shard Scaling"](advanced-cluster-new-sharding-schema.md#migration-sharded)

> ### Example1: Old Configuration

Reviewer 2: ```suggestion
### Example 1: Old Configuration (mongodbatlas_cluster) 
```

Author: Thanks! Changed in 678969f0633a27d16ba9c8891007e4bbf8b1d76d

```terraform
resource "mongodbatlas_cluster" "this" {
    project_id                   = var.project_id
    name                         = "legacy-cluster"
    cluster_type                 = "REPLICASET"

    # Provider Settings "block"
    provider_instance_size_name = "M10" # 1 Replication Spec Configuration
    provider_name               = "AWS" # 2 Provider Settings
    
    auto_scaling_disk_gb_enabled = true # 3 Auto Scaling
    cloud_backup                 = true # 4 Backup Configuration

    replication_specs {
     num_shards = 1
     regions_config {
       region_name     = "US_EAST_1"
       priority        = 7
       electable_nodes = 3 # 1 Replication Spec Configuration
       analytics_nodes = 1 # 1 Replication Spec Configuration
       read_only_nodes = 0 # 1 Replication Spec Configuration
       }
     }

    }
```

> ### Example2: New Configuration

Reviewer 2: ```suggestion
### Example 2: New Configuration (mongodbatlas_advanced_cluster) 
```

Author: Thanks! Changed in 678969f0633a27d16ba9c8891007e4bbf8b1d76d

```terraform
    resource "mongodbatlas_advanced_cluster" "this" {
    project_id             = var.project_id
    name                   = "advanced-cluster"
    cluster_type           = "REPLICASET"
    backup_enabled         = true # 4 Backup Configuration

    # Replication specs
    replication_specs {
      auto_scaling { # 3 Auto Scaling
        disk_gb_enabled = true
      }
      region_configs {
        region_name   = "US_EAST_1"
        priority    = 7
        provider_name = "AWS" # 2 Provider Settings
        
        electable_specs { # 1 Replication Spec Configuration
            instance_size = "M10"
            node_count    = 3
        }
        analytics_specs { # 1 Replication Spec Configuration
            instance_size = "M10"
            node_count    = 1
        }
      }
    }
```

### Output Changes

- container_id:
  - before: `mongodbatlas_cluster.this.replication_specs[0].container_id` was a flat string, e.g., `669644ae01bf814e3d25b963`
  - after: `mongodbatlas_advanced_cluster.this.replication_specs[0].container_id` is a map, e.g., `{"AWS:US_EAST_1": "669644ae01bf814e3d25b963"}`

> ## How to Change

Reviewer: Might this section be called "Best Practices" or something similar?

Author: Yes, updated in c164d007639ebd3edbc17acd55f525389d29423e.
WDYT?

Before doing any migration it is recommended to make a backup of your Terraform state files. (ADD_LINK_HERE)

> ### Method 1: `terraform plan -generate-config-out=cluster.tf`

Author: Not sure whether or not to enumerate different methods.
Once we have built more tooling around this.
The section description probably doesn't make much sense until we have multiple alternatives

Reviewer: I agree that it might be less clear to format the one option as a list. I'd suggest updating to a list format after additional methods have been developed / released.

Author: Agree. Updated

This method uses only Terraform native tools and are ideal for customers who:
1. Have an existing cluster without any Terraform configuration and want to manage their cluster with Terraform.
2. Have existing `mongodbatlas_cluster` resource(s) and don't want to use an external script for migrating.

#### Steps

1. Find the import IDs of the clusters you want to migrate: `{PROJECT_ID}-{CLUSTER_NAME}`, e.g., `664619d870c247237f4b86a6-legacy-cluster`
2. Add an import block per cluster to one of your `.tf` files:
  ```terraform
  import {
    to = mongodbatlas_advanced_cluster.this
    id = "664619d870c247237f4b86a6-legacy-cluster" # from step 1
  }
  ```

> 3. Run `terraform plan -generate-config-out=cluster.tf`, should generate a `cluster.tf` file and display a message similar to `Plan: 1 to import, 0 to add, 0 to change, 0 to destroy`:

Reviewer: ```suggestion
3. Run `terraform plan -generate-config-out=cluster.tf`. This should generate a `cluster.tf` file and display a message similar to `Plan: 1 to import, 0 to add, 0 to change, 0 to destroy`:
```

Author: Thanks updated in 288f7f76a58ef30cf833e0adbb1e9543b719d138

      oplog_size_mb                        = 0
      sample_refresh_interval_bi_connector = 0
      sample_size_bi_connector             = 0
      transaction_lifetime_limit_seconds   = 0
    }

    replication_specs {
      container_id = {
        "AWS:US_EAST_1" = "669644ae01bf814e3d25b963"
      }
      id         = "66978026668b7619f6f48cf2"
      zone_name  = "ZoneName managed by Terraform"

      region_configs {
        priority              = 7
        provider_name         = "AWS"
        region_name           = "US_EAST_1"

        auto_scaling {
          compute_enabled            = false
          compute_max_instance_size  = null
          compute_min_instance_size  = null
          compute_scale_down_enabled = false
          disk_gb_enabled            = false
        }

        electable_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 3
        }
        analytics_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 1
        }
      }
    }
  }
  ```

> 4. Run a `terraform apply`, you should expect to see the resource imported.

Reviewer: ```suggestion
4. Run `terraform apply`. You should see the resource imported.
```

Author: Thanks changed in 288f7f76a58ef30cf833e0adbb1e9543b719d138 and 3a97436556d18d9312d3bd84a9e5cbe6e6e6d006

> 5. Remove the "default" fields. Many fields of this resource are optional, look for fields with a `null` or `0` value or blocks you didn't specify before, e.g:

Reviewer: ```suggestion
5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value or blocks you didn't specify before, e.g:
```

Author: Thanks. Updated

   - `advanced_configuration`
   - `connection_strings`
   - `cluster_id`
   - `bi_connector_config`
6. Re-run `terraform apply` to ensure you have no plan changes.
7. Update the references from your old cluster resource: `mongodbatlas_cluster.this.XXXX` to the new `mongodbatlas_advanced_cluster.this.XXX`.

>    - Double check [output-changes](#output-changes) to ensure the meaning stays unchanged

Reviewer: ```suggestion
   - Double check [output-changes](#output-changes) to ensure the underlying configuration stays unchanged.
```

Author: Thanks! Updated in 288f7f76a58ef30cf833e0adbb1e9543b719d138

---
page_title: "Migration Guide: Cluster to Advanced Cluster"
---

# Migration Guide: Cluster to Advanced Cluster

> **Objective**: This guide explains how to replace the `mongodbatlas_cluster` resource with the `mongodbatlas_advanced_cluster` resource. The data source(s) migration only requires [output changes](#output-changes) updates as they don't manage actual clusters.

Reviewer 3: who doesn't manage actual clusters? 
who are they? This isn't clear.

> **Note**: This guide focus on the resource migration as the data source migration is only requiring a resource_type change from `data.mongodbatlas_cluster` to `data.mongodbatlas_advanced_cluster`.  However, pay attention to the [output changes.](#output-changes)

Reviewer 4: knit: in case you want to have color boxes, see: https://developer.hashicorp.com/terraform/registry/providers/docs#callouts

e.g. https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/docs/resources/advanced_cluster.md#L7

Author: Thanks for the tip!
Do you think it will help here?
Do we have any guidance on when to use/not use colors?

Author: try out: `/guides/advanced-cluster-new-sharding-schema#migration-sharded`

Key differences in the configuration:

> 1. Replication Spec Configuration: Support different node types (electable, analytics, read_only) where hardware configuration can differ between node types.

Reviewer 4: ```suggestion
1. Replication Spec Configuration: Supports different node types (electable, analytics, read_only) where hardware configuration can differ between node types.
```

Author: Thanks! done in 8667e8e5ec54e1320aaef3b41f82e970cb8a7cb8

2. Provider Settings: Moved from the top level to the replication spec allowing you to create multi-cloud clusters.
3. Auto Scaling: Moved from the top level to the replication spec allowing you to scale replication specs individually.

> 4. Backup Configuration: Renamed from `cloud_backup` to `backup_enabled`.

Reviewer 4: do we also want to comment about the new ISS schema, like moving num_shards or disk size? or at least reference the other guide?

Author: Yes! Added a link in 8667e8e5ec54e1320aaef3b41f82e970cb8a7cb8
However, uncertain how to link correctly. Do you know how?

### Example1: Old Configuration

```terraform
resource "mongodbatlas_cluster" "this" {
    project_id                   = var.project_id
    name                         = "legacy-cluster"
    cluster_type                 = "REPLICASET"

    # Provider Settings "block"

>     provider_instance_size_name = "M10" # 1

Reviewer 4: i'd add something else in the comment so you don't need to go up and down to see what they were, like:
```
# 1 rep spec
# 2 provider settings
...
```

Author: Thanks! Done in 8667e8e5ec54e1320aaef3b41f82e970cb8a7cb8

    cloud_backup                 = true # 4

    replication_specs {
     num_shards = 1
     regions_config {
       region_name     = "US_EAST_1"
       priority        = 7
       electable_nodes = 3 # 1
       analytics_nodes = 1 # 1
       read_only_nodes = 0 # 1
       }
     }

    }
```

### Example2: New Configuration

```terraform
    resource "mongodbatlas_advanced_cluster" "this" {
    project_id             = var.project_id
    name                   = "advanced-cluster"
    cluster_type           = "REPLICASET"
    backup_enabled         = true # 4

    # Replication specs
    replication_specs {
      auto_scaling { # 3
        disk_gb_enabled = true
      }
      region_configs {
        region_name   = "US_EAST_1"
        priority    = 7
        provider_name = "AWS" # 2
        
        electable_specs { # 1
            instance_size = "M10"
            node_count    = 3
        }
        analytics_specs { # 1
            instance_size = "M10"
            node_count    = 1
        }
      }
    }
```

### Output Changes

- container_id:
  - before: `mongodbatlas_cluster.this.replication_specs[0].container_id` was a flat string, e.g., `669644ae01bf814e3d25b963`

>   - now: `mongodbatlas_advanced_cluster.this.replication_specs[0].container_id` is a map, e.g., `{"AWS:US_EAST_1": "669644ae01bf814e3d25b963"}`

Reviewer 4: stringfied JSON object instead of a map?

Author: What change do you mean?

> ## Best Practices Before Migrating

Reviewer 3: Should the period be inside the link? I am not familiar with this syntax. Just checking because I've seen it a few times in this document. In rendered text, the period should be after the link.

Before doing any migration create a backup of your [Terraform state file.](https://developer.hashicorp.com/terraform/cli/commands/state)

> ## Migration using `terraform plan -generate-config-out=adv_cluster.tf`

Reviewer 5: Should we clarify in what terraform version this flag is supported? And maybe reference terraform docs on the functionality being used of generate-config-out.

Author: Great idea. Added link in dfba6241117b49b4eb7fdbe4ba55957d85ac3e74

      sample_refresh_interval_bi_connector = 0
      sample_size_bi_connector             = 0
      transaction_lifetime_limit_seconds   = 0
    }

    replication_specs {
      container_id = {
        "AWS:US_EAST_1" = "669644ae01bf814e3d25b963"
      }
      id         = "66978026668b7619f6f48cf2"
      zone_name  = "ZoneName managed by Terraform"

      region_configs {
        priority              = 7
        provider_name         = "AWS"
        region_name           = "US_EAST_1"

        auto_scaling {
          compute_enabled            = false
          compute_max_instance_size  = null
          compute_min_instance_size  = null
          compute_scale_down_enabled = false
          disk_gb_enabled            = false
        }

        electable_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 3
        }
        analytics_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 1
        }
      }
    }
  }
  ```
4. Run a `terraform apply`, you should expect to see the resource imported

> 5f. Remove the "default" fields. Many fields of this resource are optional, look for fields with a `null` value or blocks you didn't specify before, e.g:

Reviewer 4: ```suggestion
5f. Remove the "default" fields. Many fields of this resource are optional, look for fields with a `null` or `0` value or blocks you didn't specify before, e.g:
```

Author: Thanks. Done in 8667e8e5ec54e1320aaef3b41f82e970cb8a7cb8

   - `advanced_configuration`
   - `connection_strings`
   - `cluster_id`
   - `bi_connector_config`
6. Re-run `terraform apply` to ensure you have no plan changes
7. Update the references from your old cluster resource: `mongodbatlas_cluster.this.XXXX` to the new `mongodbatlas_advanced_cluster.this.XXX`
   - Double check [output-changes](#output-changes) to ensure the meaning stays unchanged
8. Replace your existing clusters with the ones from `cluster.tf` and run `terraform state rm mongodbatlas_cluster.this`
9. Re-run `terraform apply` to ensure you have no plan changes

### Method 2: `python ./migrate_cluster_to_adv_cluster.py` (No working implementation yet, but maybe by 24th of July)
This method uses a python script to migrate the cluster resources it has three checkpoints:

> 1. Confirm the updated config is ok

Reviewer 4: knit: should we end all bullet-point sentences with period - . ?

Author: Thanks done in 8667e8e5ec54e1320aaef3b41f82e970cb8a7cb8

---

> page_title: "Migration - Cluster to Advanced Cluster"

Reviewer 3: Did you mean: Migrate from Cluster to the Advanced Cluster Configuration"?

---

# Migration - Cluster to Advanced Cluster

> **Objective**: Guide users to replace the `mongodbatlas_cluster` resource with the `mongodbatlas_advanced_cluster` resource.

Reviewer 3: This guide explains how to replace the ... with the ....

> -> **NOTE:** This guide focuses on the resource migration, as the data source migration only requires a `resource_type` change from `data.mongodbatlas_cluster` to `data.mongodbatlas_advanced_cluster`.  However, pay attention to the [output changes.](#output-changes)

Reviewer 3: Do we need a note? We can just have this as part of objective. 

Also, not clear why I should pay attention to the output changes. If you mention it, explain why.

Author: Moved to Objective and updated. WDYT?

## Main Changes Between `mongodbatlas_cluster` and `mongodbatlas_advanced_cluster`

1. Replication Spec Configuration: Supports different node types (electable, analytics, read_only) where hardware configuration can differ between node types.
2. Provider Settings: Moved from the top level to the replication spec allowing you to create multi-cloud clusters.
3. Auto Scaling: Moved from the top level to the replication spec allowing you to scale replication specs individually.
4. Backup Configuration: Renamed from `cloud_backup` to `backup_enabled`.

> 5. See also ["Migration to new sharding schema and leveraging Independent Shard Scaling"](advanced-cluster-new-sharding-schema#migration-sharded)

Reviewer 3: Change the guide's title in the link, as discussed earlier

  replication_specs {
    num_shards = 1
    regions_config {
      region_name     = "US_EAST_1"
      priority        = 7
      electable_nodes = 3 # 1 Replication Spec Configuration
      analytics_nodes = 1 # 1 Replication Spec Configuration
      read_only_nodes = 0 # 1 Replication Spec Configuration
    }
  }
}
```

### Example 2: New Configuration (`mongodbatlas_advanced_cluster`)

```terraform
resource "mongodbatlas_advanced_cluster" "this" {
  project_id     = var.project_id
  name           = "advanced-cluster"
  cluster_type   = "REPLICASET"
  backup_enabled = true # 4 Backup Configuration

  replication_specs {
    region_configs {
      auto_scaling { # 3 Auto Scaling
        disk_gb_enabled = true
      }
      region_name   = "US_EAST_1"
      priority      = 7
      provider_name = "AWS" # 2 Provider Settings

      electable_specs { # 1 Replication Spec Configuration
        instance_size = "M10"
        node_count    = 3
      }
      analytics_specs { # 1 Replication Spec Configuration
        instance_size = "M10"
        node_count    = 1
      }
    }
  }
}
```

### Output Changes

- `container_id`:

>   - before: `mongodbatlas_cluster.this.replication_specs[0].container_id` was a flat string, e.g., `669644ae01bf814e3d25b963`

Reviewer 3: Shoud these start with capital letters? Before, After, Tip. 

Also, tip is not parallel to before and after. I would add Tip as a simple sentence, without Tip. I would place it closest to the area where the tip's stuff is discussed. So, move from here to somewhere else. It seem like it is too late to tell the users about this, after you've shown all of the above. The tip might need to go to the section in the overview.

Author: It is related to the `container_id`, so I prefer to keep it there. Applied the other changes ✅

  - after: `mongodbatlas_advanced_cluster.this.replication_specs[0].container_id` is a map, e.g., `{"AWS:US_EAST_1": "669644ae01bf814e3d25b963"}`
  - tip: If you have a single region you can access the `container_id` directly with: `one(values(mongodbatlas_advanced_cluster.this.replication_specs[0].container_id))`

## Best Practices Before Migrating

> Before doing any migration it is recommended to make a backup of your [Terraform state file.](https://developer.hashicorp.com/terraform/cli/commands/state)

Reviewer 3: Do not use " it is recommended". Use "create a backup of "

## Migration using `terraform plan -generate-config-out=adv_cluster.tf`

> This method uses only Terraform native tools and is ideal for customers who:

Reviewer 3: don't use "customers", use: "is ideal if you:"

> This method uses only Terraform native tools and are ideal for customers who:

Reviewer 4: ```suggestion
This method uses only Terraform native tools and is ideal for customers who:
```

> 1. Have an existing cluster without any Terraform configuration and want to manage their cluster with Terraform.

Reviewer 3: your cluster with Terraform.

2. Have existing `mongodbatlas_cluster` resource(s) and don't want to use an external script for migrating.

> ### Steps

Reviewer 3: Procedure

      transaction_lifetime_limit_seconds   = 0
    }

    replication_specs {
      container_id = {
        "AWS:US_EAST_1" = "669644ae01bf814e3d25b963"
      }
      id         = "66978026668b7619f6f48cf2"
      zone_name  = "ZoneName managed by Terraform"

      region_configs {
        priority              = 7
        provider_name         = "AWS"
        region_name           = "US_EAST_1"

        auto_scaling {
          compute_enabled            = false
          compute_max_instance_size  = null
          compute_min_instance_size  = null
          compute_scale_down_enabled = false
          disk_gb_enabled            = false
        }

        electable_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 3
        }
        analytics_specs {
          disk_iops       = 3000
          ebs_volume_type = null
          instance_size   = "M10"
          node_count      = 1
        }
      }
    }
  }
  ```
4. Run `terraform apply`. You should see the resource(s) imported: `Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.`

> 5. Remove the "default" fields. Many fields of this resource are optional. Look for fields with a `null` or `0` value or blocks you didn't specify before, e.g:

Reviewer 3: Here and elsewhere: search and replace all "e.g" with "such as", or "for example"

   - `advanced_configuration`
   - `connection_strings`
   - `cluster_id`
   - `bi_connector_config`

> 6. Re-use existing [Terraform expressions](https://developer.hashicorp.com/terraform/language/expressions). All fields in the generated config will have static values. Look in your old configuration for:

Reviewer 3: configuration

>    - variables, e.g., `var.project_id`

Reviewer 3: replace e.g.

>    - Terraform keywords, e.g., `for_each`, `count`, `depends_on`, etc.

Reviewer 3: Replace "e.g" and remove "etc".

> 7. Re-run `terraform apply` to ensure you have no plan changes: `No changes. Your infrastructure matches the configuration.`

Reviewer 3: planned

> 8. Update the references from your old cluster resource: `mongodbatlas_cluster.this.XXXX` to the new `mongodbatlas_advanced_cluster.this.XXX`.

Reviewer 3: For instances of "old" consider searching for them and replacing with "previous"

   - Double check [output-changes](#output-changes) to ensure the underlying configuration stays unchanged.
9.  Replace your existing clusters with the ones from `adv_cluster.tf` and run `terraform state rm mongodbatlas_cluster.this`.

>    - Without this step, Terraform will create a plan to destroy your existing cluster.

Reviewer 3: Remove the dash before the word Without. It is just a sentence that should follow the previous sentence.
Replace the word "destroy" with "delete".

10. Remove the import block created in step 2.

> 11. Re-run `terraform apply` to ensure you have no plan changes: `No changes. Your infrastructure matches the configuration.`

Reviewer 3: planned

> ### Explanation

Reviewer 3: Consider using: Terraform Actions

> Using the `project_id` and `cluster.name` Terraform imports your cluster and uses the new `mongodbatlas_advanced_cluster` schema to dump a configuration file. This dump includes all configurable values in the schema, but none of the configuration defined for your `mongodbatlas_cluster`. Therefore, it will likely be a lot more verbose and contain none of your original [Terraform expressions.](https://developer.hashicorp.com/terraform/language/expressions)

Reviewer 3: Add a comma, as follows:

Using the `project_id` and `cluster.name`, Terraform ...

Reviewer 3: Therefore, it will likely be a lot more verbose --- what will be verbose? The dump?
Did you mean the dump file? Use the words "the dump file" here and in the prev sentence.
Also it seems that Terraform loads the configuration file, or updates the configuration file with the new settings from the schema. Is there a particular reason for us to use the word dump here? (unless this is the terminology used in such cases in Terraform specifically).
I'd use this instead: "updates the configuration file with the new schema settings".

Author: Updated. WDYT?
